src/native/s3_client.c

```diff
@@ -0,0 +1,389 @@
+    struct aws_s3_client *client = (struct aws_s3_client *)jni_s3_client;
+
+    struct s3_client_make_meta_request_callback_data *callback_data =
+        aws_mem_acquire(allocator, sizeof(struct s3_client_make_meta_request_callback_data));
+
+    AWS_FATAL_ASSERT(callback_data);
+
+    jint jvmresult = (*env)->GetJavaVM(env, &callback_data->jvm);
+    (void)jvmresult;
+    AWS_FATAL_ASSERT(jvmresult == 0);
+
+    callback_data->java_s3_meta_request = (*env)->NewGlobalRef(env, java_s3_meta_request_jobject);
+    callback_data->java_s3_meta_request_response_handler_native_adapter =
+        (*env)->NewGlobalRef(env, java_response_handler_jobject);
+
+    AWS_FATAL_ASSERT(callback_data->java_s3_meta_request != NULL);
+
+    callback_data->java_s3_meta_request_response_handler_native_adapter =
+        (*env)->NewGlobalRef(env, java_response_handler_jobject);
+
+    AWS_FATAL_ASSERT(callback_data->java_s3_meta_request_response_handler_native_adapter != NULL);
+
+    struct aws_http_message *request_message = aws_http_message_new_request(allocator);
+    if (request_message == NULL) {
+        /* TODO */
+    }
+
+    if (aws_apply_java_http_request_changes_to_native_request(
+            env, jni_marshalled_message_data, jni_http_request_body_stream, request_message)) {
+        /* TODO */
+    }
+
+    struct aws_s3_meta_request_options meta_request_options = {
+        .type = meta_request_type,
+        .message = request_message,
+        .user_data = callback_data,
+        .headers_callback = s_on_s3_meta_request_headers_callback,
+        .body_callback = s_on_s3_meta_request_body_callback,
+        .finish_callback = s_on_s3_meta_request_finish_callback,
+        .shutdown_callback = s_on_s3_meta_request_shutdown_complete_callback};
+
+    struct aws_s3_meta_request *meta_request = aws_s3_client_make_meta_request(client, &meta_request_options);
+
+    aws_http_message_release(request_message);
+
+    if (!meta_request) {
+        aws_jni_throw_runtime_exception(
+            env, "S3Client.aws_s3_client_make_meta_request: creating aws_s3_meta_request failed");
+    }
+
+    return (jlong)meta_request;
+}
+
+static void s_on_s3_meta_request_shutdown_complete_callback(void *user_data) {
+
+    struct s3_client_make_meta_request_callback_data *callback_data =
+        (struct s3_client_make_meta_request_callback_data *)user_data;
+    JNIEnv *env = aws_jni_get_thread_env(callback_data->jvm);
+
+    if (callback_data->java_s3_meta_request != NULL) {
+        (*env)->CallVoidMethod(env, callback_data->java_s3_meta_request, s3_meta_request_properties.onShutdownComplete);
+
+        if (aws_jni_check_and_clear_exception(env)) {
+            AWS_LOGF_ERROR(
+                AWS_LS_S3_META_REQUEST,
+                "id=%p: Ignored Exception from S3MetaRequest.onShutdownCompete callback",
+                (void *)callback_data->java_s3_meta_request);
+        }
+    }
+
+    // We're done with this callback data, free it.
```

Comment:
I'd move this section into its own function so that it can be called from error handling in the constructor above.